---
title: "Tarspot_analysis"
author: "Mitch Roth"
date: "11/19/2018"
output: html_document
---

First, install and load packages that may be needed.
##Install and Load Packages
```{r}
#source('http://bioconductor.org/biocLite.R')
#biocLite('phyloseq')
#packageVersion('phyloseq')
#install.packages("ape")
#install.packages("ecodist")
#install.packages("vegan")
#install.packages("ggplot2")
#install.packages("indicspecies")
#install.packages("ggpubr")
#install.packages("gridExtra")
#install.packages("devtools")
library(phyloseq)
library(ape)
library(ecodist)
library(vegan)
library(ggplot2)
library(indicspecies)
library(ggpubr)
library(gridExtra)
library(devtools)
```

Now, let's import the files we need
##Import files and convert to phyloseq
```{r, results="hide"}
theme_set(theme_bw())
#input our OTU table
tarspot_matrix = read.table("/Users/rothmitc/odrive/MSU\ Google\ Drive/MSU/Classes/PLP\ 847\ Advanced\ Mycology/Group\ Project/data_files/otu_table_ITS_UPARSE_R1.txt", header = T)
tarspot_matrix
```

The matrix didn't seem to be imported in the correct format, so change the row and column names to a format that makes it compatible with phyloseq.

```{r, results="hide"}
#add row and column names to OTU table
tarspot_matrix <- tarspot_matrix[order(tarspot_matrix$ID),]
rownames(tarspot_matrix) <- tarspot_matrix$ID
tarspot_matrix <- tarspot_matrix[,-1]
tarspot_matrix
```

Now, repeat with importing the taxonomy table and formatting to phyloseq compatible format

```{r, results="hide"}
#input our taxonomy table
tarspot_tax = read.table("/Users/rothmitc/odrive/MSU\ Google\ Drive/MSU/Classes/PLP\ 847\ Advanced\ Mycology/Group\ Project/data_files/consensus_taxonomy_phyllachora_Msorghi.txt", sep = "\t", header = T)
tarspot_tax <- tarspot_tax[order(tarspot_tax$OTU),]
rownames(tarspot_tax) <- tarspot_tax$OTU
tarspot_tax <- tarspot_tax[,-1]
#Turns out this needs to be a "matrix" not a "data.frame" for phyloseq
tarspot_tax <- as.matrix(tarspot_tax)
```

Repeat for the master mapping file / metatdata file

```{r, results="hide"}
#Upload metadata file
tarspot_map = read.table("/Users/rothmitc/odrive/MSU\ Google\ Drive/MSU/Classes/PLP\ 847\ Advanced\ Mycology/Group\ Project/data_files/TNF_map.txt", sep = "\t", header = T)
tarspot_map <- tarspot_map[order(tarspot_map$Description),]
tarspot_meta <- data.frame(tarspot_map$plate_number,tarspot_map$plant_status,tarspot_map$Description)
colnames(tarspot_meta) <- c("plate_number", "plant_status", "Description")
rownames(tarspot_meta) <- tarspot_meta$Description
``` 

Finally, convert these files into phyloseq objects

```{r, results="hide"}
#Turn OTU table and taxonomy table into phyloseq objects
OTU <- otu_table(tarspot_matrix, taxa_are_rows = TRUE)
TAX <- tax_table(tarspot_tax)
META <- sample_data(tarspot_meta)
#Merge them all into one phyloseq file
tarspot_physeq <- phyloseq(OTU,TAX,META)
nrow(tarspot_physeq@tax_table)
nrow(tarspot_physeq@tax_table[tarspot_physeq@tax_table[,2]==""])
nrow(tarspot_physeq@tax_table[tarspot_physeq@tax_table[,2]!=""])
```

Great, so it looks like we have a phyloseq object that contains:
735 taxa in 47 samples
47 samples with 3 sample variables
735 taxa by 7 taxonomic ranks
And, at least 608 have taxonomy assigned to Phylum level



##Pruning
Not all of the taxa will be informative, and some may be outliers and cause problems in the analysis. So, we'll filter / prune them out of the full data set for some analyses.

```{r, results="hide"}
tarspot_physeq
#Filter out OTUS that don't appear > 5 times in at least 5% of the samples
tarspot_filtered <- genefilter_sample(tarspot_physeq, filterfun_sample(function(x) x > 5), A=.05*nsamples(tarspot_physeq))
tarspot_pruned <- prune_taxa(tarspot_filtered, tarspot_physeq)
tarspot_pruned@otu_table
#Transform reads into relative abundance
tarspot_pruned <- transform_sample_counts(tarspot_pruned, function(x) x/sum(x))
tarspot_pruned@otu_table
#Can also transform reads via rarification, but produces similar results throughout
#tarspot_pruned_rare <- rarefy_even_depth(tarspot_physeq, sample.size = min(sample_sums(tarspot_physeq)))
tarspot_pruned
nrow(tarspot_pruned@tax_table)
nrow(tarspot_pruned@tax_table[tarspot_pruned@tax_table[,2]==""])
nrow(tarspot_pruned@tax_table[tarspot_pruned@tax_table[,2]!=""])
```

So, removing OTUs low abundance (< 5) and are in fewer than 5% of samples reduces the number of taxa in our data set from 735 to 218. While this seems like a large reduction, I think it's a reasonably relaxed filtering strategy. Let's start exploring the data.

To make sure we had enough sequencing coverage, get some ROC curves
```{r}
ROC <- t(tarspot_physeq@otu_table)

S <- specnumber(ROC) # observed number of species
raremax <- min(rowSums(ROC))
Srare <- rarefy(ROC, raremax)
#plot(S, Srare, xlab = "Observed No. of Species", ylab = "Rarefied No. of Species")
#abline(0, 1)
rarecurve(ROC, step = 20, sample = raremax, col = "blue", cex = 0.6)
```

It looks like we have sufficient sequencing depth. Our sample with the lowest reads still had 33 thousand reads. At that sequencing depth, we still ID'ed around 75 species minimum, per sample. Increasing reads would have helped ID more, but the samples are plateauing at 33 thousand reads, indicating sufficient sequencing depth. 

##Alpha Diversity
Let's check the alpha diversity of species within each lesion type using Shannon's index

```{r}
theme_set(theme_bw())
p <- plot_richness(tarspot_physeq, x="plant_status", measures="Shannon") +
  geom_boxplot()
p
```

Using the full dataset (tarspot_physeq) and the pruned dataset (tarspot_pruned) yield very similar plots. They show that the Control and tar spot samples seem to have higher diversity measurements. Perhaps the lack of diversity among Fish eye samples is due to a higher abundance of necrotic tissues that make it harder for most species to survive and thrive here.

Since the "no ethanol" controls match their respective lesion types, merge them in with their lesion typs and re-run the diversity plot

```{r}
alpha <- tarspot_pruned
alpha@sam_data$plant_status <- as.character(alpha@sam_data$plant_status)
alpha@sam_data$plant_status[alpha@sam_data$plant_status=="Tar spot no ethanol"] <- "Tar spot"
alpha@sam_data$plant_status[alpha@sam_data$plant_status=="Fish eye no ethanol"] <- "Fish eye"
alpha@sam_data$plant_status[alpha@sam_data$plant_status=="Control"] <- "No symptoms"
alpha@sam_data$plant_status <- as.factor(alpha@sam_data$plant_status)
#Per reviewer request, drop "no symptoms" since they are uneven sample numbers
alpha@sam_data <- alpha@sam_data[alpha@sam_data$plant_status!="No symptoms",]
alpha@otu_table <- alpha@otu_table[,-c(45:47)]
#alpha@sam_data$plant_status <- relevel(alpha@sam_data$plant_status, "No symptoms")
p <- plot_richness(alpha, x="plant_status", measures="Shannon") +
  geom_boxplot()
p
```

Again, the full dataset and the pruned dataset yield very similar plots. The Shannon index still shows interesting differences between Fish eye symptoms and the others, so let's take a look at "richness" and "evenness", which are both parts of the Shannon index calculation.
This idea / code is largely taken from: https://github.com/joey711/phyloseq/issues/945

```{r}
tarspot_physeq_disease <- tarspot_physeq
tarspot_physeq_disease@otu_table <- tarspot_physeq_disease@otu_table[,-c(45:47)]
alpha_diversity <- estimate_richness(tarspot_physeq_disease, measure = c("Shannon", "Observed"))
H <- alpha_diversity$Shannon
S1 <- alpha_diversity$Observed
S <- log(S1)
evenness <- H/S
alpha_diversity$Evenness = evenness
#Again, merge the "no ethanol" controls with their respective symptoms
tarspot_meta$plant_status <- as.character(tarspot_meta$plant_status)
tarspot_meta$plant_status[tarspot_meta$plant_status=="Tar spot no ethanol"] <- "Tar spot"
tarspot_meta$plant_status[tarspot_meta$plant_status=="Fish eye no ethanol"] <- "Fish eye"
tarspot_meta$plant_status[tarspot_meta$plant_status=="Control"] <- "No symptoms"
tarspot_meta$plant_status <- as.factor(tarspot_meta$plant_status)
#Per reviewer request, drop "no symptoms" since they are uneven sample numbers
tarspot_meta <- tarspot_meta[tarspot_meta$plant_status!="No symptoms",]
tarspot_meta$plant_status <- droplevels(tarspot_meta$plant_status, "No symptoms")

#Make sure alpha_diversity file is in same order as the meta file so that it can be appended
alpha_diversity <- alpha_diversity[order(row.names(alpha_diversity)),]
#Sanity check
identical(rownames(alpha_diversity), rownames(tarspot_meta))
#Append the shannon data onto the metadata file
tarspot_meta$Shannon <- alpha_diversity$Shannon
tarspot_meta$Observed <- alpha_diversity$Observed
tarspot_meta$Evenness <- alpha_diversity$Evenness
PY <- levels(tarspot_meta$plant_status)
PY.pairs <- combn(seq_along(PY), 2, simplify = FALSE, FUN = function(i)PY[i])

pS <- ggplot(data=tarspot_meta, aes_string(x='plant_status', y='Shannon')) + 
  geom_boxplot() + 
  ylab("Shannon Index") + xlab("Lesion Type") +
  theme(axis.title.x = element_text(face="bold", size = 20),
        axis.title.y = element_text(face="bold", size = 20),
        axis.text = element_text(face="bold", size = 15)) +
  stat_compare_means(comparisons = PY.pairs, label = "p.signif") + 
  stat_compare_means(label.y = 0.8, size = 6)

pO <- ggplot(data=tarspot_meta, aes_string(x='plant_status', y='Observed')) + 
  geom_boxplot() + 
  ylab("Observed Richness") + xlab("Lesion Type") +
  theme(axis.title.x = element_text(face="bold", size = 20),
        axis.title.y = element_text(face="bold", size = 20),
        axis.text = element_text(face="bold", size = 15)) +
  stat_compare_means(comparisons = PY.pairs, label = "p.signif") + 
  stat_compare_means(label.y = 110, size = 6)

pE <- ggplot(data=tarspot_meta, aes_string(x='plant_status', y='Evenness')) + 
  geom_boxplot() + 
  ylab("Evenness") + xlab("Lesion Type") +
  theme(axis.title.x = element_text(face="bold", size = 20),
        axis.title.y = element_text(face="bold", size = 20),
        axis.text = element_text(face="bold", size = 15)) +
  stat_compare_means(comparisons = PY.pairs, label = "p.signif",
                     label.y.npc = "bottom") + 
  stat_compare_means(label.y = 0.175, size = 6)

grid.arrange(pS, pO, pE, ncol = 3)
```



So, overall it seems like the Fish eye lesions are associated with lower observed taxa (lower richness) and lower evenness. This indicates that Fish eye lesions are likely dominated by a single taxa, or small group of unique taxa, compared to Tar spot lesions and Healthy tissues. If that's the case, we can likely identify candidate taxa that are dominating in the Fish eye lesions. 

Now let's look at taxa associated with each sample. At this point, there are far too many OTU's in the full dataset (tarspot_physeq), so use the pruned dataset (tarspot_pruned), which is also normalized to the sample with the lowest read depth.

```{r}
#Bar charts by taxonomy level
#Merge the "no ethanol" controls in with their lesion type
pooled_pruned <- tarspot_pruned
pooled_pruned@sam_data$plant_status <- as.character(pooled_pruned@sam_data$plant_status)
pooled_pruned@sam_data$plant_status[pooled_pruned@sam_data$plant_status=="Tar spot no ethanol"] <- "Tar spot"
pooled_pruned@sam_data$plant_status[pooled_pruned@sam_data$plant_status=="Fish eye no ethanol"] <- "Fish eye"
pooled_pruned@sam_data$plant_status[pooled_pruned@sam_data$plant_status=="Control"] <- "No symptoms"
pooled_pruned@sam_data$plant_status <- as.factor(pooled_pruned@sam_data$plant_status)
pooled_pruned@sam_data$plant_status <- relevel(pooled_pruned@sam_data$plant_status, "No symptoms")

#merge samples
pooled_pruned_merge <- merge_samples(pooled_pruned, "plant_status")
sample_data(pooled_pruned_merge)$plant_status <- factor(sample_names(pooled_pruned_merge))
#taxa merge
pooled_pruned_merge <- tax_glom(pooled_pruned_merge, "Genus")
##Transform to percentage of total available
pooled_pruned_merge = transform_sample_counts(pooled_pruned_merge, function(x) 100 * x/sum(x))
#Relevel plant_status
pooled_pruned_merge@sam_data$plant_status <- relevel(pooled_pruned_merge@sam_data$plant_status,"No symptoms")
#generate stacked bar plot
cl <- colors(distinct = TRUE)
set.seed(44)
mycols <- sample(cl,60)
p1 <- plot_bar(pooled_pruned_merge, fill = 'Genus', x = 'plant_status') +
  xlab("Lesion Type") +
  theme(axis.title.x = element_text(face="bold", size = 20),
        axis.text.x = element_text(face="bold", size=15, angle = 270),
        axis.title.y = element_text(face="bold", size = 20),
        axis.text.y = element_text(size = 10),
        legend.title = element_text(face="bold", size = 10),
        legend.text = element_text(face="bold", size = 10),
        legend.key.size = unit(0.5, 'lines'),
        legend.position = "right") +
  scale_y_continuous(labels=function(n){format(n, scientific = FALSE)})+
  geom_bar(aes(fill = Genus ), stat = 'identity', position = 'stack') +
  scale_fill_manual(values= mycols) +
  ylab("Relative abundance (% of total sequences)")

p1

```

Interestingly, it looks like Phyllachora is present in all lesioned samples, but most abundant in Fish eye lesions, and ~less abundant in Tar spot lesions. Pool all of the samples by lesion type and plot this again.

Let's see if the communities for each type of lesion are different by running a PERMANOVA on the distance matrix, as a function of lesion type. 

```{r}
tarspot_physeq@sam_data$plant_status <- as.character(tarspot_physeq@sam_data$plant_status)
tarspot_physeq@sam_data$plant_status[tarspot_physeq@sam_data$plant_status=="Tar spot no ethanol"] <- "Tar spot"
tarspot_physeq@sam_data$plant_status[tarspot_physeq@sam_data$plant_status=="Fish eye no ethanol"] <- "Fish eye"
#tarspot_physeq@sam_data$plant_status[tarspot_physeq@sam_data$plant_status=="Control"] <- "No symptoms"
tarspot_physeq@sam_data$plant_status <- as.factor(tarspot_physeq@sam_data$plant_status)
#Per reviewer request, drop "no symptoms" since they are uneven sample numbers
tarspot_physeq <- subset_samples(tarspot_physeq,
                                 plant_status != "Control")
#tarspot_physeq@sam_data <- tarspot_physeq@sam_data[tarspot_physeq@sam_data$plant_status!="Control"]
#tarspot_physeq@otu_table <- tarspot_physeq@otu_table[,-c(45:47)]
#tarspot_physeq@sam_data$plant_status <- relevel(tarspot_physeq@sam_data$plant_status, "No symptoms")
#Calculate distance matrix
tarspot_dist <- phyloseq::distance(tarspot_physeq, method = "bray")
#Create a list of lesion types
lesion_type <- tarspot_physeq@sam_data$plant_status
#Run PERMNAOVA
p.hoc1 <- adonis(tarspot_dist ~ lesion_type)
#Post hoc test for PERMANOVA
p.hoc2 <- betadisper(tarspot_dist, lesion_type)
#TukeyHSD(p.hoc1)
TukeyHSD(p.hoc2)
```

Great! Lesion types have significantly different communities according to PERMANOVA (adonis). The betadisper function can be used to assess beta diversity, and the Tukey test indicates that the Fish eye community is significantly different from the Tar spot community. (Note, results are very similar even with the pruned data set)



##Ordination plots
Since this is a large dataset, it helps to look at the data with ordination methods, calculated by different distance metrics. So, let's generate a PCoA plot to visualize these communities. Let's use bray-curtis distances, as these are commonly used in ecological data. Since we'll want to explore taxa associated with each lesion type, we'll have to use the pruned dataset, or else the plots get too messy.

```{r}
#Per reviewer request, drop "no symptoms"
pooled_pruned <- subset_samples(pooled_pruned,
                                 plant_status != "No symptoms")
#Calculate distance matrix
tarspot_ord_bray <- ordinate(pooled_pruned, "PCoA", "bray")
#PCoA plot

p2 <- plot_ordination(pooled_pruned, tarspot_ord_bray, 
                     type="sample", color="plant_status", shape="plant_status") + 
  geom_point(size=5) + 
  stat_ellipse(level = 0.95) + 
  scale_shape_discrete(name="Lesion Type") + 
  scale_color_discrete(name="Lesion Type") + 
  theme(axis.title.x = element_text(face="bold", size = 20),
        axis.title.y = element_text(face="bold", size = 20),
        axis.text = element_text(face="bold", size = 15),
        legend.title = element_text(face="bold", size = 20),
        legend.text = element_text(face="bold", size = 20),
        legend.key.size = unit(2, 'lines'))
  
p2


```

Great, it looks like Fish eye and Tar spot separate into unique groups based on their fungal communities, which is what we expect after the PERMANOVA results above. Interestingly, the Healthy control samples are somewhere between the two lesion types, but according to the TukeyHSD test above, it is not significantly different from either lesion type. Let's see if we can identify specific taxa associated with each lesion type, again using the pruned dataset.

```{r, Bray}
tarspot_ord_bray <- ordinate(pooled_pruned, "PCoA", "bray")
#pooled_pruned@sam_data$plant_status <- ordered(pooled_pruned@sam_data$plant_status, 
#                           levels = c("Healthy", "Fish eye", "Tar spot", "Taxa"))
pOTU <- plot_ordination(pooled_pruned, tarspot_ord_bray, 
                     type="taxa", 
                     color="Phylum",
                     justDF = T)
pOTU$OTU <- rownames(pOTU)

for (i in seq_along(pOTU$OTU)){
  if (pOTU[i,8] == "Phyllachora"){
    pOTU[i,11] <- c("Phyllachora")
  }
#  else if (pOTU[i,8] == "Monographella"){
#    pOTU[i,11] <- c("Microdochium")
#  }
#  else if (pOTU[i,8] == "Microdochium"){
#    pOTU[i,11] <- "Microdochium"
#  }
  else {
    pOTU[i,11] <- "NA"
  }
}

pSample <- plot_ordination(pooled_pruned, tarspot_ord_bray, 
                     type="sample", 
                     shape="plant_status",
                     color="plant_status",
                     justDF = T)

  

#pSample$plant_status <- ordered(pSample$plant_status, 
#                           levels = c("Healthy", "Fish eye", "Tar spot"))

p4 <- ggplot() + 
  geom_point(data = pOTU, size = 3, aes(x=Axis.1, y=Axis.2, shape = V11)) + 
  geom_point(data = pSample, size = 5,
             aes(x=Axis.1, y=Axis.2, 
                color = plant_status)) +
  stat_ellipse(data = pSample, aes(x=Axis.1, y=Axis.2, color = plant_status),
               level = 0.95) +
  scale_shape_discrete(name="Lesion Type") + 
  scale_color_discrete(name="Lesion Type") + 
  xlab("Axis 1 [45.1%]") + ylab("Axis 2 [21.1%]") +
  theme(axis.title.x = element_text(face="bold", size = 20),
        axis.title.y = element_text(face="bold", size = 20),
        axis.text = element_text(face="bold", size = 15),
        #legend.title = element_text(face="bold", size = 20),
        #legend.text = element_text(face="bold", size = 20),
        #legend.key.size = unit(2, 'lines')
        legend.position=""
        )
#p4

p5 <- ggplot() + 
  geom_point(data = pOTU, color = "white", aes(x=Axis.1, y=Axis.2)) + 
  geom_point(data = pSample, size = 5, color = "white",
             aes(x=Axis.1, y=Axis.2)) +
  stat_ellipse(data = pSample, aes(x=Axis.1, y=Axis.2, color = plant_status),
               level = 0.95) +
  scale_shape_discrete(name="Lesion Type") + 
  scale_color_discrete(name="Lesion Type") + 
  geom_text(data = pOTU, aes(x=Axis.1, y=Axis.2, label=OTU, color = Genus)) +
  xlab("Axis 1 [41.5%]") + ylab("Axis 2 [18.7%]") +
  #xlim(-0.4,0.4) +
  #ylim(-0.2,0.4) +
  theme(axis.title.x = element_text(face="bold", size = 20),
        axis.title.y = element_text(face="bold", size = 20),
        axis.text = element_text(face="bold", size = 15),
        #legend.title = element_text(face="bold", size = 20),
        #legend.text = element_text(face="bold", size = 20),
        #legend.key.size = unit(2, 'lines')
        legend.position="none"
        )

#p5
grid.arrange(p4,p5, ncol = 2)
```
```{r, Jaccard}
tarspot_ord_jac <- ordinate(pooled_pruned, "PCoA", "jaccard")
pooled_pruned@sam_data$plant_status <- ordered(pooled_pruned@sam_data$plant_status, 
                           levels = c("Healthy", "Fish eye", "Tar spot", "Taxa"))
pOTU_jac <- plot_ordination(pooled_pruned, tarspot_ord_jac, 
                     type="taxa", 
                     color="Phylum",
                     justDF = T)
pOTU_jac$OTU <- rownames(pOTU_jac)
pSample_jac <- plot_ordination(pooled_pruned, tarspot_ord_jac, 
                     type="sample", 
                     shape="plant_status",
                     color="plant_status",
                     justDF = T)
pSample_jac$plant_status <- ordered(pSample_jac$plant_status, 
                           levels = c("Healthy", "Fish eye", "Tar spot"))
for (i in seq_along(pOTU_jac$OTU)){
  if (pOTU_jac[i,8] == "Phyllachora"){
    pOTU_jac[i,11] <- c("Phyllachora")
  }
#  else if (pOTU_jac[i,8] == "Monographella"){
#    pOTU_jac[i,11] <- c("Microdochium")
#  }
#  else if (pOTU_jac[i,8] == "Microdochium"){
#    pOTU_jac[i,11] <- "Microdochium"
#  }
  else {
    pOTU_jac[i,11] <- "NA"
  }
}

#pOTU_jac$Axis.1 <- pOTU_jac$Axis.1 * -1
#pSample_jac$Axis.1 <- pSample_jac$Axis.1 * -1
p6 <- ggplot() + 
  geom_point(data = pOTU_jac, size = 3, aes(x=Axis.1, y=Axis.2, shape = V11)) + 
  geom_point(data = pSample_jac, size = 5,
             aes(x=Axis.1, y=Axis.2, 
                color = plant_status)) +
  stat_ellipse(data = pSample_jac, aes(x=Axis.1, y=Axis.2, color = plant_status),
               level = 0.95) +
  scale_shape_discrete(name="Lesion Type") + 
  scale_color_discrete(name="Lesion Type") + 
  xlab("Axis 1 [32.9%]") + ylab("Axis 2 [17.0%]") +
  theme(axis.title.x = element_text(face="bold", size = 20),
        axis.title.y = element_text(face="bold", size = 20),
        axis.text = element_text(face="bold", size = 15),
        #legend.title = element_text(face="bold", size = 20),
        #legend.text = element_text(face="bold", size = 20),
        #legend.key.size = unit(2, 'lines')
        legend.position="none"
        )
#p6

p7 <- ggplot() + 
  geom_point(data = pOTU_jac, color = "white", aes(x=Axis.1, y=Axis.2)) + 
  geom_point(data = pSample_jac, size = 5, color = "white",
             aes(x=Axis.1, y=Axis.2)) +
  stat_ellipse(data = pSample_jac, aes(x=Axis.1, y=Axis.2, color = plant_status),
               level = 0.95) +
  scale_shape_discrete(name="Lesion Type") + 
  scale_color_discrete(name="Lesion Type") + 
  geom_text(data = pOTU_jac, aes(x=Axis.1, y=Axis.2, label=OTU, color = Genus)) +
  xlab("Axis 1 [30.1%]") + ylab("Axis 2 [15.4%]") +
  #xlim(-0.2,0.2) +
  #ylim(0.,0.3) +
  theme(axis.title.x = element_text(face="bold", size = 20),
        axis.title.y = element_text(face="bold", size = 20),
        axis.text = element_text(face="bold", size = 15),
        #legend.title = element_text(face="bold", size = 20),
        #legend.text = element_text(face="bold", size = 20),
        #legend.key.size = unit(2, 'lines')
        legend.position="none")

#p7
grid.arrange(p6,p7, ncol = 2)
grid.arrange(p4,p6, ncol = 2)
grid.arrange(p5,p7, ncol = 2)
```
Neat! So now we can see that there are specific OTUs associated with Fish eye lesions, perhaps driving the symptoms we see. Let's find ou what some of these are. Also, Bray and Jaccard results are almost identical, suggesting that the separation of lesion types are driven by presence/absence data, but abundance may still be playing a role.

```{r}
#Use this code to find ID of candidate OTUs by replacing "OTU_" with the OTU of interest
pOTU[pOTU$OTU=="OTU_214",]
tarspot_physeq@tax_table[rownames(tarspot_physeq@tax_table)=="OTU_2",]
```

###Indicator species analysis
Indicator species analysis looks for significant associations with taxa / OTUs and a group. In our case, we'll group by lesion type, and we'll use the whole dataset instead of the pruned one. 
Since this indicator species analysis uses multipatt, which uses permutations, different results can be obtained each time you run the analysis. I've done it enough times to show that the same 9 or so OTUs are consistently identified as significant (p < 0.05). However, not every OTU is significant every time, so I'll run 100 iterations of the analysis. I'll combine all significant indicator OTUs from each iteration into a data frame and output the table, then count the number of times each OTU was significant (n / 100), and calculate an average p value.
```{r}
#Per reviewer request, dropped "control" samples from analysis
#This caused problems in indicator species analysis, as some OTUs were ONLY identified in control samples. Therefore, during indicator species analysis, these OTU's get NA values rather than 0 or 1.
#Manually identified that there are 23 OTUs that were ONLY identified in control samples. 
#Manually remove them prior to indicator species analysis
cut_taxa <- c("OTU_294","OTU_352","OTU_472","OTU_490","OTU_244",
              "OTU_523","OTU_645","OTU_476","OTU_666","OTU_621",
              "OTU_384","OTU_622","OTU_380","OTU_561","OTU_605",
              "OTU_575","OTU_452","OTU_703","OTU_534","OTU_546",
              "OTU_578","OTU_712","OTU_554")
keep_taxa <- NULL
for(i in 1:nrow(tarspot_physeq@otu_table)){
  if(as.character(rownames(tarspot_physeq@otu_table[i])) %in% cut_taxa){}
  else{
    keep_taxa[i] <- as.character(rownames(tarspot_physeq@otu_table[i]))
  }
}
#sanity check
'OTU_1' %in% keep_taxa    #TRUE
'OTU_294' %in% keep_taxa  #FALSE

#Prune these taxa out
tarspot_physeq_ind <- prune_taxa(keep_taxa, tarspot_physeq)
'OTU_1' %in% rownames(tarspot_physeq_ind@otu_table)    #TRUE
'OTU_294' %in% rownames(tarspot_physeq_ind@otu_table)  #FALSE

#Normalize reads
tarspot_physeq@otu_table
tarspot_physeq_norm <- transform_sample_counts(tarspot_physeq_ind, function(x) x/sum(x))
tarspot_physeq_norm@otu_table

##############################################################
# WARNING - 100 iterations will take some time, and significant
# computaitonal power. I can run this on my laptop, but it's
# not happy by the end of it. Recommended changing 100 to 10
# if you're just testing the code
##############################################################
indicator.tar <- indicspecies::multipatt(as.data.frame(t(tarspot_physeq_norm@otu_table)), cluster = tarspot_physeq_norm@sam_data$plant_status, duleg = T, control = how(nperm=999)) 
summary(indicator.tar, alpha = 0.05)
```
I got 10 OTUs that were significant

Let's see what the pruned data file tells us for indicator species

```{r}
pooled_pruned@sam_data$plant_status
pooled_pruned_ind <- pooled_pruned
pooled_pruned_ind <- subset_samples(pooled_pruned,
                                plant_status != "Healthy")
pooled_pruned_ind <- subset_samples(pooled_pruned,
                                plant_status != "Taxa")
for(i in 1:nrow(pooled_pruned_ind@otu_table)){
  if(as.character(rownames(pooled_pruned_ind@otu_table[i])) %in% cut_taxa){}
  else{
    keep_taxa[i] <- as.character(rownames(pooled_pruned_ind@otu_table[i]))
  }
}

#Prune these taxa out
pooled_pruned_ind <- prune_taxa(keep_taxa, pooled_pruned_ind)
'OTU_1' %in% rownames(pooled_pruned_ind@otu_table)    #TRUE
'OTU_294' %in% rownames(pooled_pruned_ind@otu_table)  #FALSE


indicator.tar.prune <- indicspecies::multipatt(as.data.frame(t(pooled_pruned_ind@otu_table)), cluster = pooled_pruned_ind@sam_data$plant_status, duleg = T, control = how(nperm=999)) 
summary(indicator.tar.prune, alpha = 0.05)
```

I got 5 OTUs that were significant
Let's see what all of these species actually are.

```{r}
ind.otu <- as.data.frame(tarspot_physeq@tax_table)
ind.otu$otu <- rownames(ind.otu)
#Taxa in both full and pruned indicator species analysis:
indicator.tax <- ind.otu[ind.otu$otu == "OTU_2",]
indicator.tax[2,] <- ind.otu[ind.otu$otu == "OTU_6",]
indicator.tax[3,] <- ind.otu[ind.otu$otu == "OTU_7",]
indicator.tax[4,] <- ind.otu[ind.otu$otu == "OTU_33",]
indicator.tax[5,] <- ind.otu[ind.otu$otu == "OTU_262",]
indicator.tax[1:5,]
#Taxa in full indicator species analysis onlhy:
indicator.tax[7,] <- ind.otu[ind.otu$otu == "OTU_214",]
indicator.tax[8,] <- ind.otu[ind.otu$otu == "OTU_249",]
indicator.tax[9,] <- ind.otu[ind.otu$otu == "OTU_288",]
indicator.tax[10,] <- ind.otu[ind.otu$otu == "OTU_312",]
indicator.tax[11,] <- ind.otu[ind.otu$otu == "OTU_347",]
indicator.tax[12,] <- ind.otu[ind.otu$otu == "OTU_466",]
indicator.tax[7:12,]

```

It appears CONSTAX may have misclassified a few Phyllachora OTUs as "Dothidiomycetes". I took all OTUs that had a taxonomy of "Dothidiomycetes", but no further, and did a BLAST search. Of the 23 OTUs that met this criteria, 3 had a top BLAST hit to Phyllachora.

```{r}
#Find Phyllachora OTUs that might be misclassified in dothidiomycetes
#Two examples...
tarspot_physeq@tax_table[rownames(tarspot_physeq@tax_table)=="OTU_33",]
tarspot_physeq@tax_table[rownames(tarspot_physeq@tax_table)=="OTU_46",]
miss.class <- NULL
class(rownames(tarspot_physeq@tax_table[i,]))
j <- 0
for (i in 1:nrow(tarspot_physeq@tax_table)){
  if (tarspot_physeq@tax_table[i,3] == "Dothideomycetes"
      &&
      tarspot_physeq@tax_table[i,4] == ""){
    j <- j + 1
    miss.class.OTU <- rownames(tarspot_physeq@tax_table[i,])
    miss.class[j] <- miss.class.OTU
  }
}
miss.class
```

